c2rust-analyze/src/rewrite/ty.rs

```diff
@@ -0,0 +1,370 @@
+            Some(x) => x,
+            None => return ty,
+        };
+
+        assert_eq!(args.len(), 1);
+        let mut ty = args[0];
+
+        if own == Ownership::Cell {
+            ty = mk_cell(tcx, ty);
+        }
+
+        ty = match qty {
+            Quantity::Single => ty,
+            Quantity::Slice => tcx.mk_slice(ty),
+            // TODO: This should generate `OffsetPtr<T>` rather than `&[T]`, but `OffsetPtr` is NYI
+            Quantity::OffsetPtr => tcx.mk_slice(ty),
+        };
+
+        ty = match own {
+            Ownership::Raw => tcx.mk_imm_ptr(ty),
+            Ownership::RawMut => tcx.mk_mut_ptr(ty),
+            Ownership::Imm => tcx.mk_imm_ref(tcx.mk_region(ReErased), ty),
+            Ownership::Cell => tcx.mk_imm_ref(tcx.mk_region(ReErased), ty),
+            Ownership::Mut => tcx.mk_mut_ref(tcx.mk_region(ReErased), ty),
+            Ownership::Rc => todo!(),
+            Ownership::Box => todo!(),
+        };
+
+        ty
+    })
+}
+
+struct HirTyVisitor<'a, 'tcx> {
+    acx: &'a AnalysisCtxt<'a, 'tcx>,
+    rw_lcx: LabeledTyCtxt<'tcx, RewriteLabel>,
+    //mir: &'a Body<'tcx>,
+    //span_index: SpanIndex<Location>,
+    hir_rewrites: Vec<(Span, Rewrite)>,
+}
+
+impl<'a, 'tcx> HirTyVisitor<'a, 'tcx> {
+    fn handle_ty(&mut self, rw_lty: RwLTy<'tcx>, hir_ty: &hir::Ty<'tcx>) {
+        if rw_lty.label.ty_desc.is_none() && !rw_lty.label.descendant_has_rewrite {
+            // No rewrites here or in any descendant of this HIR node.
+            return;
+        }
+
+        let hir_args = match deconstruct_hir_ty(rw_lty.ty, hir_ty) {
+            Some(x) => x,
+            None => {
+                // `hir_ty` doesn't have the expected structure (for example, we expected a type
+                // like `*mut T`, but it's actually an alias `MyPtr`), so we can't rewrite inside
+                // it.  Instead, we discard it completely and pretty-print `rw_lty` (with rewrites
+                // applied).
+                let ty = mk_rewritten_ty(self.rw_lcx, rw_lty);
+                let printer = FmtPrinter::new(*self.rw_lcx, Namespace::TypeNS);
+                let s = ty.print(printer).unwrap().into_buffer();
+                self.hir_rewrites.push((hir_ty.span, Rewrite::PrintTy(s)));
+                return;
+            }
+        };
+
+        if let Some((own, qty)) = rw_lty.label.ty_desc {
+            assert_eq!(hir_args.len(), 1);
+            let mut rw = Rewrite::Sub(0, hir_args[0].span);
+
+            if own == Ownership::Cell {
+                rw = Rewrite::TyCtor("core::cell::Cell".into(), vec![rw]);
+            }
+
+            rw = match qty {
+                Quantity::Single => rw,
+                Quantity::Slice => Rewrite::TySlice(Box::new(rw)),
+                // TODO: This should generate `OffsetPtr<T>` rather than `&[T]`, but `OffsetPtr` is
+                // NYI
+                Quantity::OffsetPtr => Rewrite::TySlice(Box::new(rw)),
```

Comment:
what affect does this have on the insertion sort rewriting?

Reply:
What's `OffsetPtr`?  What is this supposed to generate?

Reply:
`OffsetPtr` is essentially pointer arithmetic, e.g. `px + size_of<type_of<x>>`. Rather than `std::ptr::offset` it appears that the address is inherently (and temporarily) calculated by getting the address of an element at a particular index instead of calculating the offset

Reply:
This causes insertion sort to use `&mut [c_int]` instead of the hypothetical `OffsetPtrMut<c_int>`.

`OffsetPtr` is a hypothetical library type that supports both forward and backward pointer arithmetic within a certain range, unlike slices which only support forward pointer arithmetic.  Whereas `&[T]` keeps a current position (`&slice[0]`) and max forward offset (`slice.len()`), `OffsetPtr<T>` would keep a current position, max forward offset, and max backward offset.  The max backward offset must be known in order to properly bounds-check pointer arithmetic with a negative offset.